src/sysupdate/updatectl.c

```diff
@@ -0,0 +1,1263 @@
+        _cleanup_(operation_freep) Operation *op = ASSERT_PTR(userdata);
+        OrderedHashmap *map = ASSERT_PTR(op->userdata);
+        const sd_bus_error *e;
+        _cleanup_free_ char *key = NULL;
+        const char *new_version, *job_path;
+        int r;
+
+        assert(reply);
+
+        e = sd_bus_message_get_error(reply);
+        if (e) {
+                r = -sd_bus_error_get_errno(e);
+
+                key = strdup(op->target_id);
+                if (!key)
+                        return log_oom();
+                r = ordered_hashmap_put(map, key, INT_TO_PTR(r));
+                if (r < 0)
+                        return log_error_errno(r, "Failed to update hashmap: %m");
+                TAKE_PTR(key);
+
+                return r;
+        }
+
+        r = sd_bus_message_read(reply, "sto", &new_version, &op->job_id, &job_path);
+        if (r < 0)
+                return bus_log_parse_error(r);
+        op->job_path = strdup(job_path);
+        if (!op->job_path)
+                return log_oom();
+        if (isempty(new_version))
+                new_version = "latest";
+
+        /* Register this job into the hashmap. This will give it a progress bar */
+        if (strchr(op->target_id, '@'))
+                key = strdup(op->target_id);
+        else
+                key = strjoin(op->target_id, "@", new_version);
+        if (!key)
+                return log_oom();
+        r = ordered_hashmap_put(map, key, INT_TO_PTR(0)); /* takes ownership of key */
+        if (r < 0)
+                return log_error_errno(r, "Failed to add target to tracking map: %m");
+        op->target_id = TAKE_PTR(key); /* just borrowing */
+
+        /* Cancel the job if the event loop exits */
+        r = sd_event_add_exit(op->event, &op->job_interrupt_source, update_interrupted, op);
+        if (r < 0)
+                return log_error_errno(r, "Failed to set up interrupt handler: %m");
+
+        /* We need to cancel the job before the final iteration of the renderer runs */
+        r = sd_event_source_set_priority(op->job_interrupt_source, SD_EVENT_PRIORITY_IMPORTANT);
+        if (r < 0)
+                return log_error_errno(r, "Failed to set interrupt priority: %m");
+
+        /* Register for progress notifications */
+        r = sd_bus_match_signal_async(
+                        op->bus,
+                        &op->job_properties_slot,
+                        bus_sysupdate_mgr->destination,
+                        job_path,
+                        "org.freedesktop.DBus.Properties",
+                        "PropertiesChanged",
+                        update_properties_changed,
+                        NULL,
+                        op);
+        if (r < 0)
+                return log_bus_error(r, NULL, op->target_id, "listen for PropertiesChanged");
+
+        /* Register for notification when the job ends */
+        r = bus_match_signal_async(
```

Comment:
Hmm so it's technically possible for this to race: job ends before we listen for the signal. Please make a follow-up commit to register for the signal in `verb_update` before we start any jobs.

I ask for a follow-up commit for easier reviewing/reworking since this is a fairly invasive change to how `updatectl` works. Then we'll squash it in